crates/wasmtime/src/module/serialization.rs

```diff
@@ -0,0 +1,715 @@
+            );
+        }
+
+        if triple.operating_system != isa.triple().operating_system {
+            bail!(
+                "Module was compiled for operating system '{}'",
+                triple.operating_system
+            );
+        }
+
+        Ok(())
+    }
+
+    fn check_flags(&self, isa: &dyn TargetIsa) -> Result<()> {
+        let host_level = isa.flags().opt_level().into();
+        if self.opt_level != host_level {
+            bail!("Module was compiled with optimization level '{:?}' but '{:?}' is expected for the host", self.opt_level, host_level);
+        }
+
+        if self.flags_hash != Self::simple_hash(isa.flags()) {
+            bail!("Module was compiled with different Cranelift flags than the host");
+        }
+
+        Ok(())
+    }
+
+    fn check_isa_flags(&self, isa: &dyn TargetIsa) -> Result<()> {
+        for flag in &self.isa_flags {
+            if !isa.is_flag_enabled(flag) {
+                bail!("Host is missing CPU flag '{}'", flag);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_strategy(&self, compiler: &Compiler) -> Result<()> {
+        #[allow(unreachable_patterns)]
+        let matches = match (self.strategy, compiler.strategy()) {
+            (CompilationStrategy::Auto, CompilationStrategy::Auto)
+            | (CompilationStrategy::Auto, CompilationStrategy::Cranelift)
+            | (CompilationStrategy::Cranelift, CompilationStrategy::Auto)
+            | (CompilationStrategy::Cranelift, CompilationStrategy::Cranelift) => true,
+            #[cfg(feature = "lightbeam")]
+            (CompilationStrategy::Lightbeam, CompilationStrategy::Lightbeam) => true,
+            _ => false,
+        };
+
+        if !matches {
+            bail!("Module was compiled with strategy '{:?}'", self.strategy);
+        }
+
+        Ok(())
+    }
+
+    fn check_int<T: Eq + std::fmt::Display>(found: T, expected: T, feature: &str) -> Result<()> {
+        if found == expected {
+            return Ok(());
+        }
+
+        bail!(
+            "Module was compiled with a {} of '{}' but '{}' is expected for the host",
+            feature,
+            found,
+            expected
+        );
+    }
+
+    fn check_bool(found: bool, expected: bool, feature: &str) -> Result<()> {
+        if found == expected {
+            return Ok(());
+        }
+
+        bail!(
+            "Module was compiled {} {} but it {} enabled for the host",
+            if found { "with" } else { "without" },
+            feature,
+            if expected { "is" } else { "is not" }
+        );
+    }
+
+    fn simple_hash<T: Hash>(v: T) -> u64 {
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        v.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn check_tunables(&self, compiler: &Compiler) -> Result<()> {
+        let other = compiler.tunables();
+
+        Self::check_int(
+            self.tunables.static_memory_bound,
```

Comment:
Could this use named fields along the lines of:

```rust
let Tunables { static_memory_bound, /* ... */ } = self.tunables;
// .. check all the fields
```

so when we add a field to `Tunables` we're sure to add a case here too?

Reply:
Excellent suggestion, I'll add that and we can then remove the assert on hashes as that was the intention for at least catching this case at runtime.